### ISSUE TITLE
Should the buildpack provide an arbitrary pre-build hook?

### ISSUE THREAD

**Issue by hunterloftis** [heroku/heroku-buildpack-nodejs#269]
For the most part, node's lifecycle scripts work well for running things before / after installs, before start, etc.

However, in some cases it would be useful to be able to modify things like .npmrc files ahead of time (since changes to those files won't be recognized from inside of the npm scripts).

**Comment by zeke** [comment]
Do you have a specific use for this? Are customers asking for it?

Does ENV override local `.npmrc`? I forget what takes precedence.

**Comment by hunterloftis** [comment]
It isn't often requested, but it has been a few times. Mostly related to configuring something about node or npm before actually running anything in node or npm.

**Comment by zeke** [comment]
cc @othiym23 and @iarna, who would know better than anyone.

**Comment by hunterloftis** [comment]
In general I'm leaning against this change, since I believe there is usually a better way to get the same result, but I'd like to hear from any buildpack users who have cases for this that I haven't seen in a ticket.

**Comment by joeybaker** [comment]
```json
```

**Comment by joeybaker** [comment]
```json

**Comment by joeybaker** [comment]
Relying on npm scripts for the deploy process has lead to several issues for us. They can best be summarized as: a deploy requires an install, but an install doesn't mean we're doing a deploy. Hooking into the `npm install` is kludgy. We've had to work around:

1. Logging into npm so that we can install private modules
2. Running a build after an install
3. Only doing any of the above in a production environment

Here's the relevant portions of our `package.json`

```json
    "set-npm-token": "if [ -n \"${NPM_TOKEN-}\" ]; then echo \"//registry.npmjs.org/:_authToken=$NPM_TOKEN\" >> .npmrc; fi",
    "preinstall": "npm run set-npm-token && if [ -n \"${DEPLOY_ENV-}\" ]; then echo installing; npm install --unsafe-perm --userconfig .npmrc --ignore-scripts; fi",
    "postinstall": "if [ -n \"${DEPLOY_ENV-}\" ]; then npm run build; fi",
```

## logging into npm
This is the trickiest bit to solve. It seems like the standard answer is to use an env var to store the login token, and use that in `.npmrc`. However, 1) this necessitates having the env var set on any machine that wants to use the repo. Not having it set throws a pretty inscrutable error. 2) our developers login into several npm accounts. It's not feasible to have a global variable on local machines. 

Solving this is hard because npm scripts can't modify the npm config and then read that config. Settings must be in place before starting npm. Our only work around is to have a script that modifies `.npmrc`, only if the env var is set, and then run `npm install` from with in the preinstall hook. It means that we're going to wind up running `npm install` twice, and had to copy the install command from the build pack (which isn't really future-proof). 

It would be great to have a hook into the deploy process to run arbitrary scripts before dependencies are installed. [CircleCI does this very well](https://circleci.com/docs/configuration#dependencies). 

## running a build step
We're using the `postinstall` hook to run a compile assets. Again, we only want this to happen after a deploy, so putting it in postinstall isn't really "correct". But more importantly, the only way to accomplish this is to set an env var that only our production servers have. This works but feels dirty. Again, it would be nice to have a hook that's outside the normal install process.

**Comment by hunterloftis** [comment]
> 1) this necessitates having the env var set on any machine that wants to use the repo. Not having it set throws a pretty inscrutable error. 2) our developers login into several npm accounts. It's not feasible to have a global variable on local machines.

Have you considered using something like local `.env` files and `node-foreman`? For instance, `npm start` can point to `nf start`, which will automatically read any local .env files (which would exist on your dev machines, and can be different per project).

> We're using the postinstall hook to run a compile assets. Again, we only want this to happen after a deploy, so putting it in postinstall isn't really "correct".

Are there scenarios in which you want to `npm install`, but you don't want to build frontend assets?

> But more importantly, the only way to accomplish this is to set an env var that only our production servers have. This works but feels dirty.

I'm not sure I agree that this is dirty; being very much in favor of 12factor, this sort of [environment-stored config](http://12factor.net/config) actually appeals to me.

**Comment by joeybaker** [comment]
I'm not sure a `.env` file would help? Even if we figure out a good way to enable creating one without checking it into the source, we're still stuck with npm trying to read a `.npmrc` file that's broken unless a `NPM_TOKEN` env var is set. Commands like `npm install` will stop working. Put another way: the problem isn't application startup, it's all the other associated scripts that are driven by npm.

To continue that train of thought, and answer your second and third questions: I almost never want to build assets when running `npm install`. Most of the time, I'm just updating packages, or installing a new package. The _only_ time I want to build assets after running an install is when deploying. 

I'm sure you know more about 12factor than I do, but my point is that the `npm install` procedure is conflated with the "deploy this app" procedure. The two things are not the same. :\

**Comment by nickpoorman** [comment]
+1 We have the same issue. A `preinstall` hook just for building looks quite messy. We have a large build pipeline that would benefit from a `pre-build` hook. Our current solution is to build everything into a directory (on our CI server), init a new git repo, and then force push it (which means we don't have matching commit IDs between heroku and github which is NOT good).

**Comment by hunterloftis** [comment]
> To continue that train of thought, and answer your second and third questions: I almost never want to build assets when running npm install. Most of the time, I'm just updating packages, or installing a new package. The only time I want to build assets after running an install is when deploying.

Could you clarify here @joeybaker? The only time `postinstall` runs is after a full `npm install` - which is what you would use to set up a project for the first time on a new machine. I'm not sure I understand the concern over installing new packages or updating, in which case `postinstall` will not run:

```
v4.1.1:~/test/express (master)$ npm install

> express@0.0.0 postinstall /Users/hloftis/test/express
> echo POSTINSTALL

POSTINSTALL

v4.1.1:~/test/express (master *)$ npm install --save jade@1.9.2
jade@1.9.2 node_modules/jade
├── commander@2.6.0
├── character-parser@1.2.1
├── void-elements@2.0.1
├── mkdirp@0.5.1 (minimist@0.0.8)
├── constantinople@3.0.2 (acorn@2.4.0)
├── with@4.0.3 (acorn@1.2.2, acorn-globals@1.0.6)
└── transformers@2.1.0 (promise@2.0.0, css@1.0.8, uglify-js@2.2.5)
```

@nickpoorman I'd be interested in more details of what necessitated that setup. What makes a preinstall hook messy?

**Comment by joeybaker** [comment]
@hunterloftis there are three additional times we run `npm install`

* when installing a new module in the project. E.g. `npm install lodash`. This isn't a part of the build process. 
* when upgrading a module in the project and the committing the change to `package.json` so other devs can get the update. E.g. `npm install --save lodash@latest`
* when pulling down another dev's code with an updated `package.json` and needing to upgrade modules. E.g.`npm install`

All of these cases run npm's install hooks.

**Comment by hunterloftis** [comment]
The last example makes sense - I don't believe the first two examples will actually run hooks:

```
v4.1.1:~/test/express (master *)$ npm install lodash@latest
lodash@3.10.1 node_modules/lodash
v4.1.1:~/test/express (master *)$ npm install lodash
lodash@3.10.1 node_modules/lodash
v4.1.1:~/test/express (master *)$ npm install

> express@0.0.0 postinstall /Users/hloftis/test/express
> echo POSTINSTALL

POSTINSTALL
v4.1.1:~/test/express (master *)$
```

One pattern I have seen on Heroku is this:

```
  "scripts": {
    "start": "bin/www",
    "postinstall": "scripts/postinstall.sh"
  },
```

Which provides an arbitrary script that can do things like check which environment it should build for. Is that something that could work for you?

By the way, thanks for taking the time to discuss this with me.

**Comment by nickpoorman** [comment]
@hunterloftis - The preinstall / postinstall hooks require a "deploy" env var to be used (all the same reasons @joeybaker gave). Our testing environment does not require the build pipeline so there is no reason to run it.

Ideally, the best solution would be to have a `prebuild` hook that we can use to build our our app. Otherwise, we would have to set up some sort of buildpack chaining that builds all of our files into a build directory, then blasts away the current directory, and then copies all the build files into that directory.

Actually, there should be two things added here. The ability to use a `prebuild` hook AND the ability to specify the directory where everything was built to, ie. `./build` and use that as the working directory for the rest of this buildpack.

**Comment by joeybaker** [comment]
@hunterloftis my bad! You're right, the hooks only run on `npm install`. But, that's still a very common occurrence. With even a small team, modules are constantly being installed and updated, so `npm install` is run a lot by devs.

The solution you describe with a env flag is our current solution, but it took a bit of figuring out to get right. It's also a hack. I'd much rather have a hook that I know only Heroku/servers will run. (Though, keep in mind: it doesn't solve the problem of needing to log into npm first.)

Again, part of what's going through my head is the ease we had configuring CircleCI as compared to the hoops we had to jump through for heroku.

**Comment by hunterloftis** [comment]
> @hunterloftis my bad! You're right, the hooks only run on npm install. But, that's still a very common occurrence. With even a small team, modules are constantly being installed and updated, so npm install is run a lot by devs.

Certainly; that totally makes sense.

> Again, part of what's going through my head is the ease we had configuring CircleCI as compared to the hoops we had to jump through for heroku.

Definitely; seeing those notes in your build system are why I opened the issue in the first place. I'd like to avoid (re)implementing a complex hook system, so what I'm considering is maybe a very simple first-run command that's just an arbitrary thing run on Heroku.

Here's a strawman proposal: `npm run heroku`.

A standard npm script run (if it exists) as the first step in the heroku build. Thoughts?

**Comment by joeybaker** [comment]
That's really close! The only thing that remains is logging in to npm as a different user. Maybe heroku code codify setting an `NPM_TOKEN` env var?

**Comment by hunterloftis** [comment]
Since `npm run heroku` would run before, and independently of, the `npm install` step, would you not be able to do the same thing you're doing right now re: .npmrc and `NPM_TOKEN`? Like `heroku` => update npm user; `install` => do installation as that user.

**Comment by joeybaker** [comment]
hmmm… yes, that could work. But, then I'd still be stuck with a `postinstall` hook that builds assets after the install and requires a special env token. You'd almost need `npm run heroku-pre-build` and `npm run heroku-post-build`. (which doesn't seem to really fit the desire for a minimal surface area).

Maybe it's more like adding another key to `package.json`?

```json
{
  "heroku": {
    "pre": ""
   , "post" ""
  }
}
```

This has the downside of not being able to as easily tests the scripts via `npm run`, but is standard~ish practice (browserify, testling, atomify).

**Comment by fiznool** [comment]
I have another use case that would benefit tremendously from this.

I have a bunch of private modules that are hosted in bitbucket repos. I've added a deploy user to each repo with read-only SSH access. I'm then using the approach [outlined here](http://stackoverflow.com/a/29677091/1171775) to add this SSH key during Heroku compilation, so that `npm install` will work correctly on Heroku.

This was working fine with npm@2, as here the `preinstall` script runs before npm attempts to fetch any of the modules in `package.json`. However, npm@3 has changed the behaviour: before the `preinstall` script is run, npm passes over every module listed in `package.json`, and recursively fetches all dependencies, in order to produce a 'maximally flat tree'. Of course, in my case this means that since the `preinstall` script has not yet run, the SSH key has not been loaded, and so when npm tries to fetch the private modules, it fails.

So, having some sort of script which can be run before `npm install` would really help to solve my problem here. As an aside, a script which runs after `npm install` would also be helpful, to perform any necessary cleanup.

**Comment by fiznool** [comment]
I got so excited I made a PR. It's working really nicely for my use case above. :smile:

**Comment by sandfox** [comment]
Not sure what the status of this issue is but anyway hopefully it might prove useful/insightful/annoying...
for post-install tasks we're using this buildpack https://github.com/sandfox/heroku-buildpack-nodejs-npm-task to run arbitary tasks such as compiling assets, pre-warming caches, sending notifications.
It's a little more complex than things mentioned above because it requires using multiple buildpacks, but we're finding the tradeoff of complexity pretty good.

**Comment by hunterloftis** [comment]
How would this strawman hold up against peoples' use cases?

package.json:
```
"scripts": {
  "heroku-prebuild": "foo",
  "heroku-postbuild": "bar"
}
```

This would seem to eliminate messing with 'isDeploy' env vars, and would also allow for stuff like ssh key creation before running `npm install`.

**Comment by hunterloftis** [comment]
(seems to be exactly what @fiznool was recommending)

**Comment by fiznool** [comment]
Unsurprisingly a +1 from me ☺ 

FWIW I've been using code from #281 for a while now in production and it's been working really nicely.

**Comment by hunterloftis** [comment]
@fiznool add tests and I'll merge the PR!

**Comment by joeybaker** [comment]
That would work for me!

**Comment by fiznool** [comment]
@hunterloftis I opened a new PR and added a test. Hopefully everything should be in order but please let me know if there's anything else I should do.

**Comment by joeybaker** [comment]
@fiznool thank you!

**Comment by busticated** [comment]
+1

i just snaked my way over here stumbling over every gotcha listed above (X_X)

a heroku-prebuild script hook would be a life-saver

**Comment by jbcpollak** [comment]
We have the same issue as @fiznool - his PR would work for us as well.

**Comment by hunterloftis** [comment]
https://devcenter.heroku.com/changelog-items/844

Done and done. Thanks everyone for your thoughts and feedback, and thanks @fiznool for the PR.

**Comment by joeybaker** [comment]
Thanks @fiznool and thanks @hunterloftis for shepherding this!
